console/collections/src/kary_merkle_tree/mod.rs

```diff
@@ -0,0 +1,271 @@
+            input.resize(ARITY as usize, empty_hash);
+
+            root_hash = path_hasher.hash_children(&input)?;
+        }
+        lap!(timer, "Hashed {} padding levels", padding_depth);
+
+        finish!(timer);
+
+        Ok(Self {
+            leaf_hasher: leaf_hasher.clone(),
+            path_hasher: path_hasher.clone(),
+            root: root_hash,
+            tree,
+            empty_hash,
+            number_of_leaves: leaves.len(),
+        })
+    }
+
+    #[inline]
+    /// Returns the Merkle path for the given leaf index and leaf.
+    pub fn prove(&self, leaf_index: usize, leaf: &LH::Leaf) -> Result<KaryMerklePath<PH, DEPTH, ARITY>> {
+        // Ensure the leaf index is valid.
+        ensure!(leaf_index < self.number_of_leaves, "The given Merkle leaf index is out of bounds");
+
+        // Compute the leaf hash.
+        let leaf_hash = self.leaf_hasher.hash_leaf(leaf)?;
+
+        // Compute the start index (on the left) for the leaf hashes level in the Merkle tree.
+        let start = match checked_next_power_of_n(self.number_of_leaves, ARITY as usize) {
+            Some(num_leaves) => (num_leaves - 1) / (ARITY as usize - 1),
+            None => bail!("Integer overflow when computing the Merkle tree start index"),
+        };
+
+        // Compute the absolute index of the leaf in the Merkle tree.
+        let mut index = start + leaf_index;
+        // Ensure the leaf index is valid.
+        ensure!(index < self.tree.len(), "The given Merkle leaf index is out of bounds");
+        // Ensure the leaf hash matches the one in the tree.
+        ensure!(self.tree[index] == leaf_hash, "The given Merkle leaf does not match the one in the Merkle tree");
+
+        // Initialize a vector for the Merkle path.
+        let mut path = Vec::with_capacity(DEPTH as usize);
+
+        // Iterate from the leaf hash to the root level, storing the sibling hashes along the path.
+        for _ in 0..DEPTH {
+            // Compute the index of the sibling hash, if it exists.
+            if let Some(siblings) = siblings::<ARITY>(index) {
+                // Append the sibling hashes to the path.
+                let sibling_hashes = siblings.iter().map(|index| self.tree[*index]).collect::<Vec<_>>();
+
+                path.push(sibling_hashes);
+                // Compute the index of the parent hash, if it exists.
+                match parent::<ARITY>(index) {
+                    // Update the index to the parent index.
+                    Some(parent) => index = parent,
+                    // If the parent does not exist, the path is complete.
+                    None => break,
+                }
+            }
+        }
+
+        // If the Merkle path length is not equal to `DEPTH`, pad the path with the empty hash.
+        let empty_hashes = (0..ARITY.saturating_sub(1)).map(|_| self.empty_hash).collect::<Vec<_>>();
+        path.resize(DEPTH as usize, empty_hashes);
```

Comment:
```suggestion
        if path.len() != DEPTH as usize {
            let empty_hashes = (0..ARITY.saturating_sub(1)).map(|_| self.empty_hash).collect::<Vec<_>>();
            path.resize(DEPTH as usize, empty_hashes);
        }
```
to avoid the `empty_hashes` alloc in case no padding is needed

Reply:
can you check if the normal merkle tree has a similar optimization possible?

Reply:
@raychu86 can you review this suggested change?

Reply:
> can you check if the normal merkle tree has a similar optimization possible?
just checked; this doesn't apply there, as we don't prepare a collection for resizing purposes there; the perf improvement here doesn't come from not actually resizing a vector (as this is basically a noop), but from not creating `empty_hashes`

Reply:
This change looks good